### ISSUE TITLE
Allow DistributedApplicationTestingBuilder to override WithLifetime(ContainerLifetime.Persistent)

### ISSUE THREAD

**Issue by yarseyah** [dotnet/aspire#6850]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

Whilst using the `DistributedApplicationTestingBuilder` for testing, when the AppHost has one or more `WithLifetime` set to persist containers, an increasing number of separate containers are generated.

The unit tests seem to generate a completely new set of resources, e.g. the first below is when running in the IDE, and the next two correspond to two test runs.

![Image](https://github.com/user-attachments/assets/d31131f3-08a8-42ac-9437-b6bb7d3f0c45)

```csharp
var fixturesDb = builder.AddRedis("fixtures")
    .WithPersistence()
    .WithDataVolume()
    .WithRedisInsight(
        cfg =>
        {
            cfg.WithLifetime(ContainerLifetime.Persistent);
        },
        "redis-insight")
    .WithLifetime(ContainerLifetime.Persistent);
```
Keeping the containers corresponding to a unit test seems to be undesirable - additionally, if I have designed the unit tests to be re-runnable on a persisted database, it would help massively not to spin up those resources again.

### Describe the solution you'd like

It would be useful to inform `DistributedApplicationTestingBuilder` how to override `ContainerLifetime.Persistent` or reuse the same containers without adding the additional 'test-run' to the resource name.

**Comment by yarseyah** [comment]
I've managed to jury-rig the following:

**AppHost**
```csharp
var containerLifetimeArgument =
    args.SingleOrDefault(a => a.StartsWith("--CONTAINER_LIFETIME="));

var containerLifetime = containerLifetimeArgument?[21..] switch
{
    "session" => ContainerLifetime.Session,
    "persistent" => ContainerLifetime.Persistent,
    _ => ContainerLifetime.Persistent,
};
```

And from the unit-test side, I can pass the `args` as such:

```csharp
 var builder = DistributedApplicationTestingBuilder
            .CreateAsync<Projects.Site_AppHost>(
            [
                "--CONTAINER_LIFETIME=session"
            ]);
```

Whilst this means that tests still create their own containers, at least they now destroy them afterwards - so I'm halfway there now.

**Comment by afscrome** [comment]
`DistributedApplicationTestingBuilder` gives you access to the resources before you fully build the app host, giving you a chance to mute them for your tests.  You should be able to do something like the following to remove any persistence.

```cs
        var appHost = await DistributedApplicationTestingBuilder.CreateAsync<Projects.Foo_AppHost>();

        foreach(var resource in appHost.Resources)
        {
            var lifetimeAnnotaitons = resource.Annotations.OfType<ContainerLifetimeAnnotation>().ToList();
            foreach (var lifetimeAnnotation in lifetimeAnnotaitons) {
                resource.Annotations.Remove(lifetimeAnnotation);
            }
        }

        await using var app = await appHost.BuildAsync();
```

**Comment by DamianEdwards** [comment]
In the Aspire samples repo I added an [extension to the tests to override the container lifetime](https://github.com/dotnet/aspire-samples/blob/0c27e4e3bac5f102db1002fd2e0e1ba894e1009a/tests/SamplesIntegrationTests/Infrastructure/DistributedApplicationExtensions.cs#L39-L53) but due to the way the `DistributedApplicationTestingBuilder` works right now you still can't share container instances between test runs or between tests and non-tests (what #6888 was covering before I closed it as a dupe of this issue).

I think we should treat the current behavior of persistent containers in `DistributedApplicationTestingBuilder` as a bug and triage it accordingly.

**Comment by akhilesh78** [comment]
var fixturesDb = builder.AddRedis("fixtures")
    .WithContainerName("redis-insight")
    .WithPersistence()
    .WithDataVolume()
    .WithRedisInsight(
        cfg =>
        {
            cfg.WithLifetime(ContainerLifetime.Persistent);
        },
        "redis-insight")
    .WithLifetime(ContainerLifetime.Persistent);

**Comment by afscrome** [comment]
> By default, if a container is marked as persistent then do not assign it a random resource name during testing. Instead, take persistence as an indicator that the container is a shared singleton instance.

The one issue I see with this is that you're making it easier for tests to "Work on my machine", by accidently relying on state.  I do wonder if we should prefix/suffix the persistent container names with a constant prefix (e.g `Test_`) so that all tests get the same persistent instance, but tests get a separate instance to your main app.

That said there is equally value in re-using the same instance as your apphost - you can quickly inspect the results of a test in pgadmin, redis commander etc as well as minimising test startup time.

> That is to automatically change the lifetimes of containers from persistent to session during testing.

There is merit to this approach as that helps guarantee your tests are fully isolated from each other.  That said, I do agree this shouldn't be done automatically - as you said it's not expected and it may also put unexpected strains on CPU / memory.  Some kind of opt in method.

I wonder if there's value in having something like `builder.WithIsolation(None | PerTest | IsolateTestsFromAppHost)`, (or perhaps an `IsolationLevel` property on `DistributedApplicationOptions`) where
- `None`: Leave persistent container resource names as in (i.e. suggestion 1)
- `IsolateTestsFromAppHost`: When run under test host, prefix persistent containers with `Test_`
- `Per Test`: The current behaviour where persistent containers are effectively treated as session containers

**Comment by DamianEdwards** [comment]
> I wonder if there's value in having something like builder.WithIsolation(None | PerTest | IsolateTestsFromAppHost), (or perhaps an IsolationLevel property on DistributedApplicationOptions) where

Yeah I think this needs to be selectable behavior from within the test, for the whole app host and/or per resource.
